launcher/logs/LogParser.cpp

```diff
@@ -0,0 +1,366 @@
+    } else {
+        m_buffer.append(data.toString());
+    }
+}
+
+std::optional<LogParser::Error> LogParser::getError()
+{
+    return m_error;
+}
+
+MessageLevel::Enum LogParser::parseLogLevel(const QString& level)
+{
+    auto test = level.trimmed().toUpper();
+    if (test == "TRACE") {
+        return MessageLevel::Trace;
+    } else if (test == "DEBUG") {
+        return MessageLevel::Debug;
+    } else if (test == "INFO") {
+        return MessageLevel::Info;
+    } else if (test == "WARN") {
+        return MessageLevel::Warning;
+    } else if (test == "ERROR") {
+        return MessageLevel::Error;
+    } else if (test == "FATAL") {
+        return MessageLevel::Fatal;
+    } else {
+        return MessageLevel::Unknown;
+    }
+}
+
+std::optional<LogParser::LogEntry> LogParser::parseAttributes()
+{
+    LogParser::LogEntry entry{
+        "",
+        MessageLevel::Info,
+    };
+    auto attributes = m_parser.attributes();
+
+    for (const auto& attr : attributes) {
+        auto name = attr.name();
+        auto value = attr.value();
+        if (name == "logger"_L1) {
+            entry.logger = value.trimmed().toString();
+        } else if (name == "timestamp"_L1) {
+            if (value.trimmed().isEmpty()) {
+                m_parser.raiseError("log4j:Event Missing required attribute: timestamp");
+                return {};
+            }
+            entry.timestamp = QDateTime::fromSecsSinceEpoch(value.trimmed().toLongLong());
+        } else if (name == "level"_L1) {
+            entry.levelText = value.trimmed().toString();
+            entry.level = parseLogLevel(entry.levelText);
+        } else if (name == "thread"_L1) {
+            entry.thread = value.trimmed().toString();
+        }
+    }
+    if (entry.logger.isEmpty()) {
+        m_parser.raiseError("log4j:Event Missing required attribute: logger");
+        return {};
+    }
+
+    return entry;
+}
+
+void LogParser::setError()
+{
+    m_error = {
+        m_parser.errorString(),
+        m_parser.error(),
+    };
+}
+
+void LogParser::clearError()
+{
+    m_error = {};  // clear previous error
+}
+
+bool isPotentialLog4JStart(QStringView buffer)
+{
+    static QString target = QStringLiteral("<log4j:event");
+    if (buffer.isEmpty() || buffer[0] != '<') {
+        return false;
+    }
+    auto bufLower = buffer.toString().toLower();
+    return target.startsWith(bufLower) || bufLower.startsWith(target);
+}
+
+std::optional<LogParser::ParsedItem> LogParser::parseNext()
+{
+    clearError();
+
+    if (m_buffer.isEmpty()) {
+        return {};
+    }
+
+    if (m_buffer.trimmed().isEmpty()) {
+        m_buffer.clear();
+        return {};
+    }
+
+    // check if we have a full xml log4j event
+    bool isCompleteLog4j = false;
+    m_parser.clear();
+    m_parser.setNamespaceProcessing(false);
+    m_parser.addData(m_buffer);
+    if (m_parser.readNextStartElement()) {
+        if (m_parser.qualifiedName() == "log4j:Event"_L1) {
```

Comment:
If you're going to make it case insensitive this should be case insensitive too, so should all attribute checks, probably.
```suggestion
        if (m_parser.qualifiedName().compare("log4j:Event"_L1, Qt::CaseInsensitive) == 0) {
```

I find it weird to accept malformed XML, though. It's gonna have no problem with 
```xml
<log4j:Event>
    ...
</log4j:EVeNt>
```